File: Lucy/2025/5-W04/17_Letter_Combinations_of_a_Phone_Number.ts
function letterCombinations(digits: string): string[] {
  if (digits.length === 0) return [];

  const digitPairMap = new Map<string, string[]>();
  digitPairMap.set("2", "abc".split(""));
  digitPairMap.set("3", "def".split(""));
  digitPairMap.set("4", "ghi".split(""));
  digitPairMap.set("5", "jkl".split(""));
  digitPairMap.set("6", "mno".split(""));
  digitPairMap.set("7", "pqrs".split(""));
  digitPairMap.set("8", "tuv".split(""));
  digitPairMap.set("9", "wxyz".split(""));

  const results: string[] = [];

  const path: string[] = [];
  const backtrack = (depth: number) => {
    if (depth === digits.length) {
      results.push(path.join(""));
      return;
    }

    for (const digit of digitPairMap.get(digits[depth])!) {
      path.push(digit);
      backtrack(depth + 1);
      path.pop();
    }
  };
  backtrack(0);

  return results;
}

File: Meli/LeetCode 75/17. Letter Combinations of a Phone Number.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        mapping = {
            "2": ["a", "b", "c"],
            "3": ["d", "e", "f"],
            "4": ["g", "h", "i"],
            "5": ["j", "k", "l"],
            "6": ["m", "n", "o"],
            "7": ["p", "q", "r", "s"],
            "8": ["t", "u", "v"],
            "9": ["w", "x", "y", "z"]
            }
        results = []
        def backtrack(digits_list, path, results):
            if len(digits_list) == 0:
                if len(path) > 0:
                    results.append("".join(path))
                return
            digit = digits_list[0]
            for char in mapping[digit]:
                path.append(char)
                if len(digits_list) > 1:
                    backtrack(digits_list[1:], path, results)
                else:
                    backtrack([], path, results)
                path.pop()
        
        backtrack(list(digits), [], results)
        return results